moirae/models/thevenin/__init__.py

```diff
@@ -0,0 +1,112 @@
+"""Interface to the `Thevenin package <https://github.com/ROVI-org/thevenin>`_"""
+import numpy as np
+from functools import partial
+from typing import Iterator
+
+from thevenin import Model, Experiment
+
+from ..base import CellModel, InputQuantities, OutputQuantities
+from .state import TheveninASOH, TheveninTransient
+from .ins_outs import TheveninInput
+
+
+class TheveninModel(CellModel):
+    """
+    Connection between Moirae and an ECM model implemented via `Thevenin <https://github.com/ROVI-org/thevenin>`_.
+
+    Args:
+        isothermal: Whether to treat the system as isothermal
+    """
+
+    def __init__(self, isothermal: bool = False):
+        self.isothermal = isothermal
+        # TODO (wardlt): Add a maximum timestep size? Victor was using that to avoid the issues with the integrator
+
+    def _make_models(self, transient: TheveninTransient, asoh: TheveninASOH, inputs: TheveninInput) -> Iterator[Model]:
+        """
+        Generate a model for each member of the batch of experimental conditions
+
+        Args:
+            transient: Batch of transient states
+            asoh: Batch of health variables
+
+        Yields:
+            A model for each member of the batch
+        """
+
+        for b in range(max(transient.batch_size, asoh.batch_size, inputs.batch_size)):
+            # The value of each member of the transient or ASOH is a 2D array with the first dimension either 1
+            #  or batch_size. The % signs below are a short syntax for either using the same value for all batches
+            #  (anything mod 1 is 0) or the appropriate member of the batch
+            params = {'num_RC_pairs': asoh.num_rc_elements, 'isothermal': self.isothermal}
+            for scalar, value in [
+                ('soc0', transient.soc), ('capacity', asoh.capacity), ('mass', asoh.mass), ('Cp', asoh.c_p),
+                ('T_inf', inputs.t_inf), ('h_therm', asoh.h_thermal), ('A_therm', asoh.a_therm)
+            ]:
+                params[scalar] = value[b % value.shape[0], 0]
+
+            # Add the SOC and series resistors as functions where we pin the batch ID to the appropriate value
+            params['ocv'] = partial(asoh.ocv, batch_id=b)
+            params['R0'] = partial(asoh.r[0], batch_id=b)
+
+            # Append the RC elements
+            for r in range(params['num_RC_pairs']):
+                params[f'R{r + 1}'] = partial(asoh.r[r + 1], batch_id=b)
+                params[f'C{r + 1}'] = partial(asoh.r[r], batch_id=b)

```

Comment:
Interesting. Did you make your `lambda soc, ...` function inside a for loop that also involved making `r_copy`? If so, the value of `r_copy` will be the last one created in the loop because the code inside a lambda is evaluated [when the code is called](https://docs.python.org/3/glossary.html#term-lambda):

```
>>> f = lambda x: a + 2  # a is not defined, but that's OK
>>> a = 2  
>>> f(1)  # Because the value isn't evaluated until here
4
>>>
```

Partial functions, on the other hand, hold on to the object that was passed during creation (so it must exist)

```
>>> a = 4
>>> f = partial(np.add, a)
>>> a = 2
>>> f(2)
6
```

As an aside, the above example is a good reason to avoid assigning Lambda functions to a variable. They're really only meant to be temporary, as also visible in that Python won't serialize them.

TL;DR: I think `lambda` is the problem, but am going to add a test anyway

Reply:
Ah, I see, I think your hunch may be correct, as I was defining the `lambda`s inside a `for`-loop. At any rate, when writing the tests, I would encourage us to test a case with more than one RC element in the ECM.

Reply:
Done. Caught that I was using the resistance for both R and C while at it 🤦🏼‍♂️

Reply:
sweet! code twice, test thrice 💡